src/main/java/de/tum/in/www1/artemis/service/connectors/localci/buildagent/LocalCIDockerService.java

```diff
@@ -246,15 +255,103 @@ public void pullDockerImage(LocalCIBuildJobQueueItem buildJob, BuildLogsMap buil
 
     @Scheduled(cron = "${artemis.continuous-integration.image-cleanup.cleanup-schedule-time:0 0 3 * * *}")
     public void deleteOldDockerImages() {
-
         if (!imageCleanupEnabled) {
             log.info("Docker image cleanup is disabled");
             return;
         }
 
+        Set<String> imageNames = getUnusedDockerImages();
+
         // Get map of docker images and their last build dates
         IMap<String, ZonedDateTime> dockerImageCleanupInfo = hazelcastInstance.getMap("dockerImageCleanupInfo");
 
+        // Delete images that have not been used for more than imageExpiryDays days
+        for (String dockerImage : dockerImageCleanupInfo.keySet()) {
+            if (imageNames.contains(dockerImage)) {
+                if (dockerImageCleanupInfo.get(dockerImage).isBefore(ZonedDateTime.now().minusDays(imageExpiryDays))) {
+                    log.info("Deleting docker image {}", dockerImage);
+                    try {
+                        dockerClient.removeImageCmd(dockerImage).exec();
+                    }
+                    catch (NotFoundException e) {
+                        log.warn("Docker image {} not found during cleanup", dockerImage);
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Checks for available disk space and triggers the cleanup of old Docker images if the available space falls below a specified threshold.
+     * The threshold is configurable via the 'artemis.continuous-integration.image-cleanup.disk-space-threshold-mb' property in the application settings.
+     *
+     * @implNote - We use the Docker root directory to check disk space availability. This is in case the Docker images are stored on a separate partition.
+     *           - We need to iterate over the map entries since don't remove the oldest image from the map.
+     */
+
+    @Scheduled(fixedRateString = "${artemis.continuous-integration.image-cleanup.disk-space-check-interval-ms:3600000}", initialDelayString = "${artemis.continuous-integration.image-cleanup.disk-space-check-interval-ms:3600000}")
+    public void checkUsableDiskSpaceThenCleanUp() {
+        if (imageCleanupEnabled) {
+            try {
+                // Get the Docker root directory to check disk space.
+                File file = new File(Objects.requireNonNullElse(dockerClient.infoCmd().exec().getDockerRootDir(), "/"));
```

Comment:
I stumbled over this because I was confused in line 331, because I first thought in every loop it would only be checked if the newly cleared image would be larger than 2000mb. 
A better name for the variable would have prevented that confusion for me. Besides my confusion I do not think that `file` is a descriptive variable name.

```suggestion
                File dockerRootDirectory = new File(Objects.requireNonNullElse(dockerClient.infoCmd().exec().getDockerRootDir(), "/"));
```